metabase/metabase
# Allow removing filter defaults on Subscriptions (EE)

Closes https://github.com/metabase/metabase/issues/24223

Some background on subscription filters in OSS vs EE:
* In OSS, custom parameters are ignored— only the default dashboard params are used.
* In EE, custom parameters are merged over the dashboard defaults

In EE, any custom parameters with `null` or `undefined` values were always mapped to the dashboard’s default values on the backend (at the time each subscription notification is sent), making it impossible to clear the parameter.

This PR allows these parameters to be cleared by making the following distinction:
* `undefined` means “use dashboard’s default value”
* `null` means “don’t filter”

Note:
* EE subscription parameters are initialized as `undefined` when created in the UI.
* EE parameters with `undefined` values are variably synchronized on the backend with the dashboard’s defaults whenever those change— just like OSS subscriptions.
* If the user changes an EE parameter in the UI, it cannot be re-synchronized to the default (since clearing the parameter now sets it to `null` and we don’t have UI for setting it back to `undefined`)

A big part of the diff is just removing the prop-drilling for `defaultParametersById`.  This prop seems obsolete, because defaults are attached to parameters appropriately. Since we don’t need to recall them by ID anywhere, I did some housekeeping to remove them.

### How to verify

Start email server with

```
docker run -p 1080:1080 -p 1025:1025 maildev/maildev
```

Then try to reproduce the linked issue.  Send test email in the UI to verify that default filters are removed in the emailed results.  Check updated e2e specs for extra edge cases.

### Checklist

- [x] Wait for https://github.com/metabase/metabase/pull/33758 to be merged and removed hard-coded dashboard ID from e2e spec
